From 8664bef45fb2fa1266e41bb499da7b765e81f818 Mon Sep 17 00:00:00 2001
From: "mujiburrahimank@maxlinear.com"
 <mujiburrahimank@lvdocker7.lvsw.maxlinear.com>
Date: Thu, 18 Dec 2025 06:28:29 -0800
Subject: [PATCH] WLANRTSYS-97395 : Upstep hostap-ng to the latest opensource
 commit

Fix for segfault and interface not being up issue due to misalignment
between MXL and latest opensource
---
 src/ap/hostapd.c                   |  5 +++++
 src/ap/mxl_wpa_auth.c              |  6 ++----
 src/ap/wpa_auth_glue.c             |  4 +++-
 src/drivers/driver_nl80211_event.c | 15 +++++++++++++++
 4 files changed, 25 insertions(+), 5 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 59d74089c..c818242bd 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -2531,7 +2531,12 @@ static int hostapd_no_ir_channel_list_updated(struct hostapd_iface *iface,
 	if (!hw_features) {
 		wpa_printf(MSG_DEBUG,
 			   "Could not fetching hardware channel list");
+#ifndef CONFIG_VENDOR_MXL
 		return -1;
+#else /* CONFIG_VENDOR_MXL */
+		wpa_printf(MSG_DEBUG, "fetching hardware channel list failed / or radio disabled");
+		return 0;
+#endif /* CONFIG_VENDOR_MXL */
 	}
 
 	all_no_ir = true;
diff --git a/src/ap/mxl_wpa_auth.c b/src/ap/mxl_wpa_auth.c
index 4a7bb6803..390552fd6 100644
--- a/src/ap/mxl_wpa_auth.c
+++ b/src/ap/mxl_wpa_auth.c
@@ -312,8 +312,7 @@ void mxl_ml_wpa_ptkinitnegotiating(struct wpa_state_machine *sm)
 		/* WPA2 send GTK in the 4-way handshake */
 		secure = 1;
 		gtk_len = gsm->GTK_len;
-		if (sm->wpa_auth->conf.disable_gtk ||
-		    sm->wpa_key_mgmt == WPA_KEY_MGMT_OWE) {
+		if (sm->wpa_auth->conf.disable_gtk) {
 			/*
 			 * Provide unique random GTK to each STA to prevent use
 			 * of GTK in the BSS.
@@ -552,8 +551,7 @@ void mxl_ml_wpa_ptk_group_rekeynegotiating(struct wpa_state_machine *sm)
 			"sending 1/2 msg of Group Key Handshake");
 
 	gtk = gsm->GTK[gsm->GN - 1];
-	if (sm->wpa_auth->conf.disable_gtk ||
-	    sm->wpa_key_mgmt == WPA_KEY_MGMT_OWE) {
+	if (sm->wpa_auth->conf.disable_gtk) {
 		/*
 		 * Provide unique random GTK to each STA to prevent use
 		 * of GTK in the BSS.
diff --git a/src/ap/wpa_auth_glue.c b/src/ap/wpa_auth_glue.c
index c6b9c0e8b..116dacafb 100644
--- a/src/ap/wpa_auth_glue.c
+++ b/src/ap/wpa_auth_glue.c
@@ -581,6 +581,8 @@ static int hostapd_wpa_auth_set_key(void *ctx, int vlan_id, enum wpa_alg alg,
 	}  /* if only CONFIG_WIFI_CERTIFICATION flag is on */
 #endif /* CONFIG_TESTING_OPTIONS */
 #endif /* CONFIG_TESTING_OPTIONS || CONFIG_WIFI_CERTIFICATION */
+	return hostapd_drv_set_key(ifname, hapd, alg, addr, idx, vlan_id,
+				   1, NULL, 0, key, key_len, key_flag);
 #else /* CONFIG_VENDOR_MXL */
 #ifdef CONFIG_TESTING_OPTIONS
 	if (key_flag & KEY_FLAG_MODIFY) {
@@ -623,9 +625,9 @@ static int hostapd_wpa_auth_set_key(void *ctx, int vlan_id, enum wpa_alg alg,
 		hapd->last_gtk_len = key_len;
 	}
 #endif /* CONFIG_TESTING_OPTIONS */
-#endif /* CONFIG_VENDOR_MXL */
 	return hostapd_drv_set_key(ifname, hapd, alg, addr, idx, vlan_id,
 				   set_tx, NULL, 0, key, key_len, key_flag);
+#endif /* CONFIG_VENDOR_MXL */
 }
 
 
diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index f9cf35ab6..a3c77babd 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -4615,7 +4615,9 @@ int process_global_event(struct nl_msg *msg, void *arg)
 	u64 wdev_id = 0;
 	int wdev_id_set = 0;
 	int wiphy_idx_set = 0;
+#ifndef CONFIG_VENDOR_MXL
 	bool processed = false;
+#endif /* CONFIG_VENDOR_MXL */
 
 	/* Event marker, all prior events have been processed */
 	if (gnlh->cmd == NL80211_CMD_GET_PROTOCOL_FEATURES) {
@@ -4664,6 +4666,7 @@ int process_global_event(struct nl_msg *msg, void *arg)
 			    (wiphy_idx_set && wiphy_idx == wiphy_idx_rx) ||
 			    (wdev_id_set && bss->wdev_id_set &&
 			     wdev_id == bss->wdev_id)) {
+#ifndef CONFIG_VENDOR_MXL
 				processed = true;
 				do_process_drv_event(bss, gnlh->cmd, tb);
 				/* There are two types of events that may need
@@ -4683,10 +4686,21 @@ int process_global_event(struct nl_msg *msg, void *arg)
 				 * happened. */
 				if (!nl80211_drv_in_list(global, drv))
 					break;
+#else /* CONFIG_VENDOR_MXL */
+				do_process_drv_event(bss, gnlh->cmd, tb);
+				return NL_SKIP;
+#endif /* CONFIG_VENDOR_MXL */
 			}
 		}
+#ifdef CONFIG_VENDOR_MXL
+		wpa_printf(MSG_DEBUG,
+			   "nl80211: Ignored event %d (%s) for foreign interface (ifindex %d wdev 0x%llx wiphy %d)",
+			   gnlh->cmd, nl80211_command_to_string(gnlh->cmd),
+			   ifidx, (long long unsigned int) wdev_id, wiphy_idx_rx);
+#endif /* CONFIG_VENDOR_MXL */
 	}
 
+#ifndef CONFIG_VENDOR_MXL
 	if (processed)
 		return NL_SKIP;
 
@@ -4694,6 +4708,7 @@ int process_global_event(struct nl_msg *msg, void *arg)
 		   "nl80211: Ignored event %d (%s) for foreign interface (ifindex %d wdev 0x%llx wiphy %d)",
 		   gnlh->cmd, nl80211_command_to_string(gnlh->cmd),
 		   ifidx, (long long unsigned int) wdev_id, wiphy_idx_rx);
+#endif /* CONFIG_VENDOR_MXL */
 
 	return NL_SKIP;
 }
-- 
2.43.0

